### DISCUSSION TITLE
How to list methods not wrapped in a certain statement?

### DISCUSSION THREAD

**Discussion by markpahulje** [github/codeql#5282] [Q&A] [answered]
How would you list methods **not** in a try-catch clause?

```
var fi2 = new FileInfo(path2);
```

How to list methods **not** wrapped in Using statement?

```
StreamWriter sw = new StreamWriter("hello.txt")
```

**Comment by aeisenberg** [comment, marked as answer]
I'm assuming this is in C#. Something like this should work to find method calls not lexically contained by a try-catch.

```
import csharp

predicate inTryCatch(MethodCall mc) {
  exists(TryStmt ts | ts = mc.getParent*())
    or exists(CatchClause cc | cc = mc.getParent*())
}

from MethodCall mc
where not inTryCatch(mc)
select mc
```

Of course, this only finds method calls _lexically_ enclosed. Are you looking for method calls that are part of a data flow that is not in a try-catch block? For that you will need to use [data-flow analysis](https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-csharp/).

**Reply by markpahulje** [reply]
This will suffice, thank you.
